token/js/src/extensions/tlvData.ts

```diff
@@ -0,0 +1,95 @@
+export type TLVNumberSize = 1 | 2 | 4 | 8;
+
+function readTLVNumberSize<T>(
+    buffer: Buffer,
+    size: TLVNumberSize,
+    offset: number,
+    constructor: (x: number | bigint) => T
+): T {
+    switch (size) {
+        case 1:
+            return constructor(buffer.readUInt8(offset));
+        case 2:
+            return constructor(buffer.readUInt16LE(offset));
+        case 4:
+            return constructor(buffer.readUInt32LE(offset));
+        case 8:
+            return constructor(buffer.readBigUInt64LE(offset));
+    }
+}
+
+export class TLVData {
+    private readonly tlvData: Buffer;
+    private readonly typeSize: TLVNumberSize;
+    private readonly lengthSize: TLVNumberSize;
+
+    public constructor(buffer: Buffer, typeSize: TLVNumberSize = 2, lengthSize: TLVNumberSize = 2, offset: number = 0) {
+        this.tlvData = buffer.subarray(offset);
+        this.typeSize = typeSize;
+        this.lengthSize = lengthSize;
+    }
+
+    /**
+     * Get the raw tlv data
+     *
+     * @return the raw tlv data
+     */
+    public get data(): Buffer {
+        return this.tlvData;
+    }
+
+    /**
+     * Get a single entry from the tlv data. This function performs a greedy search returning the first entry with the given type.
+     *
+     * @param type the type of the entry to get
+     *
+     * @return the entry from the tlv data or null
+     */
+    public entry(type: Buffer | Uint8Array | number | bigint): Buffer | null {
+        let offset = 0;
+        while (offset + this.typeSize + this.lengthSize <= this.tlvData.length) {
+            let typeMatches = false;
+            switch (typeof type) {
+                case 'number':
+                    typeMatches = readTLVNumberSize(this.tlvData, this.typeSize, offset, Number) === type;
+                    break;
+                case 'bigint':
+                    typeMatches = readTLVNumberSize(this.tlvData, this.typeSize, offset, BigInt) === type;
+                    break;
+                default:
+                    typeMatches = this.tlvData.subarray(offset, offset + this.typeSize).equals(type);
+                    break;
+            }
+            offset += this.typeSize;
+            const entryLength = readTLVNumberSize(this.tlvData, this.lengthSize, offset, Number);
+            offset += this.lengthSize;
+            if (typeMatches && offset + entryLength <= this.tlvData.length) {
+                return this.tlvData.subarray(offset, offset + entryLength);
+            }
+            offset += entryLength;
+        }
+        return null;
+    }
```

Comment:
I see what you're doing here, and this function does, given a type (discriminator), return the value at that TLV entry, however I think making the distinction in the code as specifically a **discriminator** is a worthy inclusion.

Similar to length, we use specific-length arrays for discriminators. Token2022 is using `u16`s converted to bytes, while TLV is using an SPL discriminator. Maybe we can somehow include this? FWIW, I know Token2022 seems to duplicate a lot of code from TLV 😭 but the JS lib doesn't have to!

https://github.com/solana-labs/solana-program-library/blob/f35dc5fdb6fc01d054afc51a6ea5a10307a48e3b/token/program-2022/src/extension/mod.rs#L862-L866

https://github.com/solana-labs/solana-program-library/blob/f35dc5fdb6fc01d054afc51a6ea5a10307a48e3b/libraries/discriminator/src/discriminator.rs#L24-L26

Reply:
My goal was to make it so generic that you can basically supply anything as a discriminator and this helper class would then figure out how to parse / read that from bytes. I think it might make sense to add a simple version of splDiscriminate to the js library as well (more on that below)

Reply:
You think maybe define SplDiscriminate separately and then let Token2022 or some other lib drive this TLV lib with whatever size discriminator they need?

Ideally we could then re-write this whole file https://github.com/solana-labs/solana-program-library/blob/master/token/js/src/extensions/extensionType.ts